packages/@lwc/ssr-runtime/src/render.ts

```diff
@@ -39,22 +58,80 @@ export function* fallbackTmpl(
     }
 }
 
+export function fallbackTmplNoYield(
+    emit: (segment: string) => void,
+    _props: unknown,
+    _attrs: unknown,
+    _slotted: unknown,
+    Cmp: LightningElementConstructor,
+    _instance: unknown
+) {
+    if (Cmp.renderMode !== 'light') {
+        emit('<template shadowrootmode="open"></template>');
+    }
+}
+
 export type GenerateMarkupFn = (
     tagName: string,
     props: Record<string, any> | null,
     attrs: Attributes | null,
     slotted: Record<number | string, AsyncGenerator<string>> | null
 ) => AsyncGenerator<string>;
 
+export type GenerateMarkupFnAsyncNoGen = (
+    emit: (segment: string) => void,
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Attributes | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => Promise<void>;
+
+export type GenerateMarkupFnSyncNoGen = (
+    emit: (segment: string) => void,
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Attributes | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => void;
+
+type GenerateMarkupFnVariants =
+    | GenerateMarkupFn
+    | GenerateMarkupFnAsyncNoGen
+    | GenerateMarkupFnSyncNoGen;
+
 export async function serverSideRenderComponent(
     tagName: string,
-    compiledGenerateMarkup: GenerateMarkupFn,
-    props: Record<string, any>
+    compiledGenerateMarkup: GenerateMarkupFnVariants,
+    props: Record<string, any>,
+    mode: 'asyncYield' | 'async' | 'sync' = 'asyncYield'
 ): Promise<string> {
     let markup = '';
 
-    for await (const segment of compiledGenerateMarkup(tagName, props, null, null)) {
-        markup += segment;
+    if (mode === 'asyncYield') {
+        for await (const segment of (compiledGenerateMarkup as GenerateMarkupFn)(
+            tagName,
+            props,
+            null,
+            null
+        )) {
+            markup += segment;
+        }
+    } else if (mode === 'async') {
+        const emit = (segment: string) => {
+            markup += segment;
+        };
+        await (compiledGenerateMarkup as GenerateMarkupFnAsyncNoGen)(
+            emit,
+            tagName,
+            props,
+            null,
+            null
+        );
+    } else if (mode === 'sync') {
+        const emit = (segment: string) => {
+            markup += segment;
+        };
+        (compiledGenerateMarkup as GenerateMarkupFnSyncNoGen)(emit, tagName, props, null, null);
```

Comment:
This function call is synchronous and technically could be split from `serverSideRenderComponent ` as a separate exported function. But from what I recall, there weren't objections for this API being asynchronous from the consumer standpoint, and maintaining the async function signature allowed for unifying all three modes as seen above.